src/Juvix/Compiler/Core/Transformation/UnrollRecursion.hs

```diff
@@ -1,7 +1,88 @@
 module Juvix.Compiler.Core.Transformation.UnrollRecursion (unrollRecursion) where
 
+import Data.HashMap.Strict qualified as HashMap
+import Juvix.Compiler.Core.Data.IdentDependencyInfo
+import Juvix.Compiler.Core.Data.InfoTableBuilder
+import Juvix.Compiler.Core.Extra
 import Juvix.Compiler.Core.Transformation.Base
 
--- TODO: not yet implemented / at first only check for recursion and give an error
 unrollRecursion :: InfoTable -> InfoTable
-unrollRecursion tab = tab
+unrollRecursion tab =
+  let (mp, (tab', ())) =
+        run $
+          runState @(HashMap Symbol Symbol) mempty $
+            runInfoTableBuilder tab $
+              forM_ (buildSCCs (createIdentDependencyInfo tab)) goSCC
+   in mapIdentSymbols mp $ pruneInfoTable tab'
+  where
+    mapIdentSymbols :: HashMap Symbol Symbol -> InfoTable -> InfoTable
+    mapIdentSymbols mp = over infoMain adjustMain . mapAllNodes (umap go)
+      where
+        go :: Node -> Node
+        go = \case
+          NIdt idt@Ident {..} ->
+            case HashMap.lookup _identSymbol mp of
+              Just sym' -> NIdt idt {_identSymbol = sym'}
+              Nothing -> NIdt idt
+          node -> node
+
+        adjustMain :: Maybe Symbol -> Maybe Symbol
+        adjustMain = \case
+          Just sym -> Just $ fromMaybe sym (HashMap.lookup sym mp)
+          Nothing -> Nothing
+
+    goSCC :: Members '[InfoTableBuilder, State (HashMap Symbol Symbol)] r => SCC Symbol -> Sem r ()
+    goSCC = \case
+      CyclicSCC syms -> unrollSCC syms
+      AcyclicSCC _ -> return ()
+
+    unrollSCC :: Members '[InfoTableBuilder, State (HashMap Symbol Symbol)] r => [Symbol] -> Sem r ()
+    unrollSCC syms = do
+      freshSyms <- genSyms
+      forM_ syms (unroll freshSyms)
+      modify (\mp -> foldr (mapSymbol freshSyms) mp syms)
+      where
+        unrollLimit :: Int
+        unrollLimit = 140
+
+        mapSymbol :: HashMap (Symbol, Int) Symbol -> Symbol -> HashMap Symbol Symbol -> HashMap Symbol Symbol
+        mapSymbol freshSyms sym = HashMap.insert sym (fromJust $ HashMap.lookup (sym, unrollLimit) freshSyms)
+
+        genSyms :: forall r. Member InfoTableBuilder r => Sem r (HashMap (Symbol, Int) Symbol)
```

Comment:
Have you considered using `Indexed Symbol` instead of `(Symbol, Int`)?

Reply:
I wasn't aware it exists. I changed to use it.